File: sim/NationRise.Core/Military/UnitCatalogue.cs
namespace NationRise.Core.Military;

/*
   Starting roster with values taken from the unit research. Kept in code for
   now because the numbers are still moving; once they settle this becomes a
   data file the pipeline emits like everything else.
*/
public static class UnitCatalogue
{
    private static float[] Ratings(
        float infantry = 0, float armour = 0, float vehicle = 0, float helicopter = 0,
        float fixedWing = 0, float missile = 0, float ship = 0, float submarine = 0,
        float building = 0, float population = 0)
        => [infantry, armour, vehicle, helicopter, fixedWing, missile, ship, submarine, building, population];

    public static readonly UnitClass MotorizedInfantry = new()
    {
        Id = "motorized_infantry",
        Name = "Motorized Infantry",
        Domain = Domain.Land,
        Armour = ArmourClass.Infantry,
        MaxHitPoints = 15f,
        Speed = 0.90f,
        Echelon = 1,
        Attack = Ratings(infantry: 3.0f, armour: 2.0f, vehicle: 3.0f, building: 0.1f, population: 2.0f),
        Defence = Ratings(infantry: 3.8f, armour: 2.5f, vehicle: 3.8f, helicopter: 0.6f, fixedWing: 0.3f),
    };

    public static readonly UnitClass MechanizedInfantry = new()
    {
        Id = "mechanized_infantry",
        Name = "Mechanized Infantry",
        Domain = Domain.Land,
        Armour = ArmourClass.Armour,
        MaxHitPoints = 25f,
        Speed = 1.10f,
        Echelon = 1,
        Attack = Ratings(infantry: 3.0f, armour: 5.0f, vehicle: 5.0f, building: 0.2f, population: 1.5f),
        Defence = Ratings(infantry: 3.8f, armour: 6.3f, vehicle: 6.3f, helicopter: 0.9f, fixedWing: 0.4f),
    };

    public static readonly UnitClass MainBattleTank = new()
    {
        Id = "main_battle_tank",
        Name = "Main Battle Tank",
        Domain = Domain.Land,
        Armour = ArmourClass.Armour,
        MaxHitPoints = 45f,
        Speed = 1.20f,
        Echelon = 1,
        Attack = Ratings(infantry: 9.0f, armour: 8.0f, vehicle: 9.0f, building: 0.5f, population: 1.0f),
        Defence = Ratings(infantry: 9.0f, armour: 8.0f, vehicle: 9.0f, helicopter: 1.0f, fixedWing: 0.5f),
    };

    public static readonly UnitClass TowedArtillery = new()
    {
        Id = "towed_artillery",
        Name = "Towed Artillery",
        Domain = Domain.Land,
        Armour = ArmourClass.UnarmouredVehicle,
        MaxHitPoints = 12f,
        Speed = 0.60f,
        Echelon = 3,
        Attack = Ratings(infantry: 7.0f, armour: 4.0f, vehicle: 7.0f, building: 3.0f, population: 4.0f),
        Defence = Ratings(infantry: 1.5f, armour: 1.0f, vehicle: 1.5f),
    };

    public static readonly UnitClass NavalInfantry = new()
    {
        Id = "naval_infantry",
        Name = "Naval Infantry",
        Domain = Domain.Land,
        Armour = ArmourClass.Infantry,
        MaxHitPoints = 19f,
        Speed = 0.95f,
        Echelon = 1,
        Attack = Ratings(infantry: 6.0f, armour: 3.0f, vehicle: 6.0f, building: 0.2f, population: 2.0f),
        Defence = Ratings(infantry: 5.0f, armour: 3.0f, vehicle: 5.0f, helicopter: 0.8f, fixedWing: 0.4f),
    };

    public static readonly UnitClass Corvette = new()
    {
        Id = "corvette",
        Name = "Corvette",
        Domain = Domain.Sea,
        Armour = ArmourClass.SurfaceShip,
        MaxHitPoints = 30f,
        Speed = 2.20f,
        Echelon = 1,
        Attack = Ratings(ship: 6.0f, submarine: 4.0f, infantry: 2.0f, building: 1.0f),
        Defence = Ratings(ship: 6.0f, submarine: 3.0f, fixedWing: 2.5f, missile: 1.5f),
    };

    public static readonly UnitClass Destroyer = new()
    {
        Id = "destroyer",
        Name = "Destroyer",
        Domain = Domain.Sea,
        Armour = ArmourClass.SurfaceShip,
        MaxHitPoints = 45f,
        Speed = 2.60f,
        Echelon = 1,
        Attack = Ratings(ship: 9.0f, submarine: 7.0f, infantry: 3.0f, building: 2.0f),
        Defence = Ratings(ship: 9.0f, submarine: 6.0f, fixedWing: 4.0f, missile: 3.0f),
    };

    public static readonly IReadOnlyList<UnitClass> All =
    [
        MotorizedInfantry, MechanizedInfantry, MainBattleTank, TowedArtillery, NavalInfantry,
        Corvette, Destroyer,
    ];

    public static UnitClass ById(string id) =>
        All.FirstOrDefault(u => u.Id == id)
        ?? throw new KeyNotFoundException($"Unknown unit class '{id}'.");
}

File: web/src/sim/military/UnitCatalogue.ts
import { f32 } from '~/sim/determinism/float32'
import { ArmourClass, ARMOUR_CLASS_COUNT } from '~/sim/military/ArmourClass'
import { Domain, type UnitClass } from '~/sim/military/UnitClass'

/*
 * Daftar awal dengan nilai dari riset unit. Masih di dalam kode karena angkanya
 * belum berhenti bergerak; begitu mengendap ia menjadi berkas data yang
 * dikeluarkan pipeline seperti yang lain.
 */

interface Ratings {
  infantry?: number
  armour?: number
  vehicle?: number
  helicopter?: number
  fixedWing?: number
  missile?: number
  ship?: number
  submarine?: number
  building?: number
  population?: number
}

function ratings(values: Ratings): Float32Array {
  const out = new Float32Array(ARMOUR_CLASS_COUNT)

  out[ArmourClass.Infantry] = values.infantry ?? 0
  out[ArmourClass.Armour] = values.armour ?? 0
  out[ArmourClass.UnarmouredVehicle] = values.vehicle ?? 0
  out[ArmourClass.Helicopter] = values.helicopter ?? 0
  out[ArmourClass.FixedWing] = values.fixedWing ?? 0
  out[ArmourClass.Missile] = values.missile ?? 0
  out[ArmourClass.SurfaceShip] = values.ship ?? 0
  out[ArmourClass.Submarine] = values.submarine ?? 0
  out[ArmourClass.Building] = values.building ?? 0
  out[ArmourClass.Population] = values.population ?? 0

  return out
}

export const MOTORIZED_INFANTRY: UnitClass = {
  id: 'motorized_infantry',
  name: 'Motorized Infantry',
  domain: Domain.Land,
  armour: ArmourClass.Infantry,
  maxHitPoints: 15,
  speed: f32(0.9),
  echelon: 1,
  attack: ratings({ infantry: 3.0, armour: 2.0, vehicle: 3.0, building: 0.1, population: 2.0 }),
  defence: ratings({ infantry: 3.8, armour: 2.5, vehicle: 3.8, helicopter: 0.6, fixedWing: 0.3 }),
}

export const MECHANIZED_INFANTRY: UnitClass = {
  id: 'mechanized_infantry',
  name: 'Mechanized Infantry',
  domain: Domain.Land,
  armour: ArmourClass.Armour,
  maxHitPoints: 25,
  speed: f32(1.1),
  echelon: 1,
  attack: ratings({ infantry: 3.0, armour: 5.0, vehicle: 5.0, building: 0.2, population: 1.5 }),
  defence: ratings({ infantry: 3.8, armour: 6.3, vehicle: 6.3, helicopter: 0.9, fixedWing: 0.4 }),
}

export const MAIN_BATTLE_TANK: UnitClass = {
  id: 'main_battle_tank',
  name: 'Main Battle Tank',
  domain: Domain.Land,
  armour: ArmourClass.Armour,
  maxHitPoints: 45,
  speed: f32(1.2),
  echelon: 1,
  attack: ratings({ infantry: 9.0, armour: 8.0, vehicle: 9.0, building: 0.5, population: 1.0 }),
  defence: ratings({ infantry: 9.0, armour: 8.0, vehicle: 9.0, helicopter: 1.0, fixedWing: 0.5 }),
}

export const TOWED_ARTILLERY: UnitClass = {
  id: 'towed_artillery',
  name: 'Towed Artillery',
  domain: Domain.Land,
  armour: ArmourClass.UnarmouredVehicle,
  maxHitPoints: 12,
  speed: f32(0.6),
  echelon: 3,
  attack: ratings({ infantry: 7.0, armour: 4.0, vehicle: 7.0, building: 3.0, population: 4.0 }),
  defence: ratings({ infantry: 1.5, armour: 1.0, vehicle: 1.5 }),
}

export const NAVAL_INFANTRY: UnitClass = {
  id: 'naval_infantry',
  name: 'Naval Infantry',
  domain: Domain.Land,
  armour: ArmourClass.Infantry,
  maxHitPoints: 19,
  speed: f32(0.95),
  echelon: 1,
  attack: ratings({ infantry: 6.0, armour: 3.0, vehicle: 6.0, building: 0.2, population: 2.0 }),
  defence: ratings({ infantry: 5.0, armour: 3.0, vehicle: 5.0, helicopter: 0.8, fixedWing: 0.4 }),
}

export const CORVETTE: UnitClass = {
  id: 'corvette',
  name: 'Corvette',
  domain: Domain.Sea,
  armour: ArmourClass.SurfaceShip,
  maxHitPoints: 30,
  speed: f32(2.2),
  echelon: 1,
  attack: ratings({ ship: 6.0, submarine: 4.0, infantry: 2.0, building: 1.0 }),
  defence: ratings({ ship: 6.0, submarine: 3.0, fixedWing: 2.5, missile: 1.5 }),
}

export const DESTROYER: UnitClass = {
  id: 'destroyer',
  name: 'Destroyer',
  domain: Domain.Sea,
  armour: ArmourClass.SurfaceShip,
  maxHitPoints: 45,
  speed: f32(2.6),
  echelon: 1,
  attack: ratings({ ship: 9.0, submarine: 7.0, infantry: 3.0, building: 2.0 }),
  defence: ratings({ ship: 9.0, submarine: 6.0, fixedWing: 4.0, missile: 3.0 }),
}

export const ALL_UNITS: readonly UnitClass[] = [
  MOTORIZED_INFANTRY,
  MECHANIZED_INFANTRY,
  MAIN_BATTLE_TANK,
  TOWED_ARTILLERY,
  NAVAL_INFANTRY,
  CORVETTE,
  DESTROYER,
]

export function unitById(id: string): UnitClass {
  const found = ALL_UNITS.find((unit) => unit.id === id)
  if (found === undefined) {
    throw new Error(`Unknown unit class '${id}'.`)
  }

  return found
}
